File: services/core-api/src/core_api/routers/admin_orders.py
"""Маршруты staff-feed заказов (admin-orders-api, PDD §4.5, INV-010).

GET /api/v1/admin/orders               → OrderListResponse
GET /api/v1/admin/orders/{order_id}    → OrderResponse

RBAC: только {ADMIN, BARISTA}, прописано в rbac_matrix.ROUTE_MATRIX.
401/403 обрабатывает RBACMiddleware — здесь явных auth-deps нет.
"""
from __future__ import annotations

# START_MODULE_CONTRACT
#   PURPOSE: HTTP routes for the staff orders feed under /api/v1/admin/orders
#            — list and detail. Read-only views for ADMIN/BARISTA.
#   SCOPE:   Pagination + status/type filtering of orders for staff.
#            No state mutations live here (see order_actions.py for that).
#   DEPENDS: M-SHARED (enums.OrderStatus/OrderType), M-DATABASE (Session),
#            core_api.services.order_history, core_api.deps.database,
#            RBACMiddleware via rbac_matrix.ROUTE_MATRIX (auth enforced
#            before handler dispatch).
#   LINKS:   docs/development-plan.xml M-CORE-API, PDD §4.5, §6.1,
#            INV-002, INV-010 (role isolation).
#   ROLE:    RUNTIME
#   MAP_MODE: EXPORTS
# END_MODULE_CONTRACT
#
# START_MODULE_MAP
#   router                  - APIRouter("/api/v1/admin", tags=["admin-orders"])
#   list_admin_orders       - GET /api/v1/admin/orders
#   get_admin_order_detail  - GET /api/v1/admin/orders/{order_id}
# END_MODULE_MAP

import uuid

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from core_api.deps import database as _db_dep
from core_api.schemas.order_history import OrderListResponse, OrderResponse
from core_api.services.order_history import (
    OrderNotFoundForStaffError,
    get_order_for_staff,
    list_orders_for_staff,
)
from shared.enums import OrderStatus, OrderType

router = APIRouter(prefix="/api/v1/admin", tags=["admin-orders"])


# Обёртка для patch-friendly dep resolution (см. routers/order_history.py)
def _get_session():
    yield from _db_dep.get_session()


def _coerce_status(raw: str) -> OrderStatus | str:
    """Принимает 'active' или любое значение OrderStatus, иначе → 422."""
    if raw == "active":
        return "active"
    try:
        return OrderStatus(raw)
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"Invalid status filter: {raw!r}",
        ) from exc


# START_CONTRACT: list_admin_orders
#   PURPOSE: Paginated staff feed of orders, filterable by status and type.
#   INPUTS:  status (query, default "active"), type (query, OrderType|None),
#            page, per_page, Session.
#   OUTPUTS: 200 OrderListResponse; 422 invalid status filter.
#   SIDE_EFFECTS: none (read-only DB query).
#   LINKS:   PDD §4.5, §6.1, INV-002, INV-010, services.order_history.
# END_CONTRACT: list_admin_orders
@router.get("/orders", response_model=OrderListResponse)
def list_admin_orders(
    status: str = Query("active"),
    type: OrderType | None = Query(None),
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
    db: Session = Depends(_get_session),
) -> OrderListResponse:
    """Пагинированный staff-фид заказов (см. PDD §4.5, S-ADMIN-003)."""
    status_filter = _coerce_status(status)
    return list_orders_for_staff(
        status_filter=status_filter,
        type_filter=type,
        page=page,
        per_page=per_page,
        db_session=db,
    )


# START_CONTRACT: get_admin_order_detail
#   PURPOSE: Single order view for staff (no ownership check; staff sees all).
#   INPUTS:  order_id: UUID (path), Session.
#   OUTPUTS: 200 OrderResponse; 404 order_not_found.
#   SIDE_EFFECTS: none.
#   LINKS:   PDD §4.5, §6.1, INV-002, INV-010, services.order_history.
# END_CONTRACT: get_admin_order_detail
@router.get("/orders/{order_id}", response_model=OrderResponse)
def get_admin_order_detail(
    order_id: uuid.UUID,
    db: Session = Depends(_get_session),
) -> OrderResponse:
    """Одиночный заказ для staff — без ownership-check."""
    try:
        return get_order_for_staff(order_id=order_id, db_session=db)
    except OrderNotFoundForStaffError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="order_not_found",
        ) from exc

File: web/admin/src/api/admin-orders.ts
// Типы зеркалят Pydantic-схемы из services/core-api/src/core_api/schemas/order_history.py
// (admin-orders-api) и schemas/order.py (staff-transitions + cancel).
// Клиент — тонкий, без state-machine: UI решает какую кнопку показать,
// сервер решает легитимность перехода (INV-016).

import { authenticatedFetch, ApiError } from './client';

// START_MODULE_CONTRACT
//   PURPOSE: Typed client for admin-only order endpoints — list/get/transition/cancel.
//   SCOPE:   Wraps /api/v1/admin/orders/* and /api/v1/orders/{id}/{status,cancel};
//            UI decides which button to show, server enforces transition legality.
//   DEPENDS: ./client (authenticatedFetch, ApiError); mirrors core-api Pydantic schemas.
//   LINKS:   docs/development-plan.xml M-WEB-ADMIN, PDD §6.1 order state machine,
//            INV-002 (admin scope enforced server-side), INV-014 (order_items
//            include menu_item_name snapshot for historical fidelity),
//            INV-016 (state-machine transitions).
//   ROLE:    RUNTIME
//   MAP_MODE: EXPORTS
// END_MODULE_CONTRACT
//
// START_MODULE_MAP
//   ApiError                  - re-export from ./client
//   OrderStatus               - PDD §6.1 state union
//   OrderType                 - 'pickup' | 'delivery'
//   AdminOrderStatusFilter    - server statuses + 'active' aggregate + 'all'
//   OrderItemResponse         - one line of an order, name snapshot per INV-014
//   OrderResponse             - full order shape returned by GET endpoints
//   OrderListResponse         - paginated list shape
//   ListAdminOrdersParams     - query params for listAdminOrders
//   listAdminOrders           - GET /api/v1/admin/orders
//   getAdminOrder             - GET /api/v1/admin/orders/{id}
//   updateOrderStatus         - PATCH /api/v1/orders/{id}/status (INV-016)
//   cancelAdminOrder          - POST /api/v1/orders/{id}/cancel (INV-016)
// END_MODULE_MAP

export { ApiError };

// ── Enums ────────────────────────────────────────────────────────────────────

export type OrderStatus =
  | 'created'
  | 'paid'
  | 'preparing'
  | 'ready'
  | 'in_delivery'
  | 'completed'
  | 'cancelled';

export type OrderType = 'pickup' | 'delivery';

// UI-шный фильтр: 'active' и конкретный OrderStatus — серверные значения,
// 'all' — клиентский sentinel, означающий "не передавать status=" (будет
// эквивалентен дефолту сервера). В текущем цикле 'all' не используется
// таб-бэкой, но клиент умеет корректно отбросить значение.
export type AdminOrderStatusFilter = OrderStatus | 'active' | 'all';

// ── Response shapes (mirror core_api.schemas.order_history) ──────────────────

export interface OrderItemResponse {
  id: string;
  menu_item_id: number | null;
  menu_item_name_ru: string;
  menu_item_name_en: string;
  size_option_id: number | null;
  size_label: string | null;
  unit_price: number;
  modifiers_snapshot: unknown[];
  quantity: number;
  line_total: number;
}

export interface OrderResponse {
  id: string;
  user_id: string;
  status: OrderStatus;
  type: OrderType;
  subtotal: number;
  discount_amount: number;
  delivery_fee: number;
  total: number;
  created_at: string;
  items: OrderItemResponse[];
}

export interface OrderListResponse {
  orders: OrderResponse[];
  total_count: number;
  page: number;
  per_page: number;
}

// ── Request params ───────────────────────────────────────────────────────────

export interface ListAdminOrdersParams {
  status?: AdminOrderStatusFilter;
  type?: OrderType;
  page?: number;
  per_page?: number;
}

// ── helpers ──────────────────────────────────────────────────────────────────

async function json<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await authenticatedFetch(path, init);
  if (res.status === 204) return undefined as unknown as T;
  return res.json() as Promise<T>;
}

function buildListQuery(params: ListAdminOrdersParams | undefined): string {
  if (!params) return '';
  const usp = new URLSearchParams();
  // 'all' — явное "снять фильтр статуса", пропускаем параметр.
  if (params.status && params.status !== 'all') usp.set('status', params.status);
  if (params.type) usp.set('type', params.type);
  if (params.page != null) usp.set('page', String(params.page));
  if (params.per_page != null) usp.set('per_page', String(params.per_page));
  const qs = usp.toString();
  return qs ? `?${qs}` : '';
}

// ── API functions ────────────────────────────────────────────────────────────

// START_CONTRACT: listAdminOrders
//   PURPOSE: Fetch paginated admin order list with optional status/type filter.
//   INPUTS:  params?: ListAdminOrdersParams — status/type/page/per_page; status='all'
//            means omit the param so the server uses its default.
//   OUTPUTS: Promise<OrderListResponse>
//   SIDE_EFFECTS: GET request; throws ApiError on non-2xx (401 → redirect via client.ts).
//   LINKS:   INV-002 (server enforces admin/barista scope), PDD §6.1.
// END_CONTRACT: listAdminOrders
export const listAdminOrders = (
  params?: ListAdminOrdersParams,
): Promise<OrderListResponse> =>
  json(`/api/v1/admin/orders${buildListQuery(params)}`);

// START_CONTRACT: getAdminOrder
//   PURPOSE: Fetch single order by id (used to refresh after a transition).
//   INPUTS:  orderId: string — UUID of the order
//   OUTPUTS: Promise<OrderResponse>
//   SIDE_EFFECTS: GET request; throws ApiError (404 → not found).
//   LINKS:   INV-002, PDD §6.1.
// END_CONTRACT: getAdminOrder
export const getAdminOrder = (orderId: string): Promise<OrderResponse> =>
  json(`/api/v1/admin/orders/${orderId}`);

// START_CONTRACT: updateOrderStatus
//   PURPOSE: Drive the order state machine forward (paid→preparing→ready→completed).
//            Server validates the transition; client only sends the requested next state.
//   INPUTS:  orderId: string, newStatus: OrderStatus
//   OUTPUTS: Promise<OrderResponse> — updated order with new status.
//   SIDE_EFFECTS: PATCH; 409 on illegal transition, 403 on wrong role, 401 on session.
//   LINKS:   INV-016 (state-machine transitions are server-enforced; this is the only
//            client surface that triggers them via Mark Ready / Hand Out / etc).
// END_CONTRACT: updateOrderStatus
export const updateOrderStatus = (
  orderId: string,
  newStatus: OrderStatus,
): Promise<OrderResponse> =>
  json(`/api/v1/orders/${orderId}/status`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ new_status: newStatus }),
  });

// START_CONTRACT: cancelAdminOrder
//   PURPOSE: Force-cancel an order (admin only). Bypasses the linear state machine
//            but is itself a state-machine transition gated server-side.
//   INPUTS:  orderId: string, reason?: string | null — optional cancel reason text.
//   OUTPUTS: Promise<OrderResponse> — order with status='cancelled'.
//   SIDE_EFFECTS: POST; 409 if order is already finalized, 403 if not admin.
//   LINKS:   INV-002 (admin role enforced server-side), INV-016.
// END_CONTRACT: cancelAdminOrder
export const cancelAdminOrder = (
  orderId: string,
  reason?: string | null,
): Promise<OrderResponse> =>
  json(`/api/v1/orders/${orderId}/cancel`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ reason: reason ?? null }),
  });
